File: data_structures/python/btree.py
class tnode:
    def __init__(s,d,l,r,p):
        s.data = d
        s.left = l
        s.right = r
        s.parent = p
        s.size = 1

class btree:
    def __init__(s, d):
        s.root = tnode(d, None, None, None)
    def Add(s, d):
        temp = s.root
        while(True):
            temp.size+=1
            if(d<temp.data and temp.left!=None):
                temp = temp.left
            elif(d<temp.data and temp.left == None):
                t = tnode(d,None,None,temp)
                temp.left = t
                break
            elif(d>=temp.data and temp.right !=None):
                temp = temp.right
            elif(d>=temp.data and temp.right == None):
                t = tnode(d,None,None,temp)
                temp.right = t
                break
        return
    def remove(s,d):
        temp = s.root
        while(True):
            if(d==temp.data):
                break
            elif(temp.left!=None and d<temp.data):
                temp = temp.left
            elif(d>=temp.data and temp.right!=None):
                temp = temp.right
            elif(temp.left==None and temp.right==None):
                return False
        if(temp.data == d):
            if(temp.right == None and temp.left==None):
                if(temp.parent.left == temp):
                    temp.parent.left = None
                else:
                    temp.parent.right = None
            elif(temp.left!=None and temp.right==None):
                if(temp!=s.root and temp.parent.left == temp):
                    temp.parent.left = temp.left
                elif(temp!=s.root and temp.parent.right==temp):
                    temp.parent.right = temp.left
                temp.left.parent = temp.parent
                temp = None
            elif(temp.right!=None and temp.left==None):
                if(temp!=s.root and temp.parent.left==temp):
                    temp.parent.left = temp.right
                elif(temp!=s.root and temp.parent.right == temp):
                    temp.parent.right = temp.right
                temp.right.parent = temp.parent
                temp = None
            else:
                t2 = temp.right
                while(t2.left!=None):
                    t2 = t2.left
                temp.data = t2.data
                if(t2.right!=None):
                    if(t2.parent.left==t2):
                        t2.parent.left=t2.right
                    if(t2.parent.right==t2):
                        t2.parent.right=t2.right
                    t2.right.parent = t2.parent
                else:
                    if(t2.parent.left == t2):
                        t2.parent.left=None
                    else:
                        t2.parent.right = None
                    while(t2.parent!=None):
                        t2 = t2.parent
                        t2.size -=1
                    t2=None
                    temp = None
                    return True
            while(temp.parent!=None):
                temp = temp.parent
                temp.size-=1
            temp = None
            return True
        return False
    def Find(s,d):
        temp = s.root
        while(temp.data!=d):
            if(temp.data>d):
                if(temp.left !=None):
                    temp = temp.left
                else:
                    return None
            else:
                if(temp.right!=None):
                    temp = temp.right
                else:
                    return None
        return temp
    def preorder(s,t):#note root must be passed as argument eg - preorder(tree.root)
        if(t==None):
            return
        else:
            print(t.data, end=',')
            s.preorder(t.left)
            s.preorder(t.right)
    def postorder(s,t):
        if(t==None):
            return
        else:
            s.postorder(t.left)
            s.postorder(t.right)
            print(t.data, end=',')
    def inorder(s,t):
        if(t==None):
            return
        else:
            s.inorder(t.left)
            print(t.data, end=',')
            s.inorder(t.right)
    def levelorder(s,t):
        v = []
        v.append(t)
        w = []
        while(len(v) !=0):
            temp  = v[0]
            if(temp.left!=None):
                v.append(temp.left)
            if(temp.right!=None):
                v.append(temp.right)
            w.append(temp.data)
            del v[0]
        n = len(w)
        for i in range(n):
            print(w[i],end=',')
    def size(s):
        return s.root.size
        

File: data_structures/java/btree.java
import java.util.*;
class tnode{
    tnode left;
    tnode right;
    tnode parent;
    int data;
    int size;
    tnode(int d, tnode left, tnode rit, tnode parent){
        this.data = d;
        this.left = left;
        this.right = rit;
        this.parent = parent;
        size = 1;
    }
}

public class btree {
	tnode root;
	btree(int d){
		root = new tnode(d,null,null,null);
	}
	void add(int d) {//if the root of the subtree is equal to d then it is put to right
			tnode temp=root;
			while(true) {
                temp.size++;
				if(d<temp.data && temp.left!=null) {
					temp=temp.left;
				}else if(d<temp.data && temp.left==null) {
					tnode t=new tnode(d,null,null,temp);
					temp.left=t;
					break;
				}else if(d>=temp.data && temp.right!=null) {
					temp=temp.right;
				}else if(d>=temp.data && temp.right==null) {
					tnode t=new tnode(d,null,null,temp);
					temp.right=t;
					break;
				}
			}
			
	}
    boolean remove(int d){
        tnode temp = root;
        while(true){
            if(d == temp.data){
                break;
            }
            else if(temp.left!=null && d<temp.data){
                temp = temp.left;
            }else if(d>=temp.data && temp.right != null){
                temp = temp.right;
            }else if(temp.left==null && temp.right==null){
                return false;
            }
        }
        if(temp.data == d){
			if(temp.right==null && temp.left==null){
				if(temp.parent.left==temp){temp.parent.left=null;}
				else{temp.parent.right = null;}
			}else if(temp.left!=null && temp.right==null){
				if(temp!=root && temp.parent.left==temp){
                    temp.parent.left=temp.left;
                }else if(temp!=root && temp.parent.right==temp){
                    temp.parent.right=temp.left;
                }
                temp.left.parent=temp.parent;
				temp = null;
			}else if(temp.right!=null && temp.left==null){
				if(temp!=root && temp.parent.left==temp){
                    temp.parent.left=temp.right;
                }else if(temp!=root && temp.parent.right==temp){
                    temp.parent.right=temp.right;
                }
                temp.right.parent=temp.parent;
				temp = null;
			}else{
				tnode t2 = temp.right;
				while(t2.left!=null){t2 = t2.left;}	//finding the successor and replacing it with temp
				temp.data = t2.data;
				if(t2.right!=null){					//delecting the successor node
					if(t2.parent.left==t2){
						t2.parent.left=t2.right;
					}else if(t2.parent.right==t2){
						t2.parent.right=t2.right;
					}
					t2.right.parent=t2.parent;
				}else{
					if(t2.parent.left==t2){ t2.parent.left=null;}
					else{t2.parent.right = null;}
				}
				while(t2.parent!=null){t2=t2.parent; t2.size--;}
				t2 = null; temp=null;
				return true;
			}
			while(temp.parent!=null){temp = temp.parent; temp.size--;}
			temp=null;
			return true;
        }
        return false;
    }
	tnode find(int d){		//finds the tnode with data == d, least depth
		tnode temp = root;
		while(temp.data!=d){
			if(temp.data>d){
				if(temp.left!=null){temp = temp.left;}
				else{return null;}
			}else{
				if(temp.right!=null){temp=temp.right;}
				else{return null;}
			}
		}
		return temp;
	}
	void preorder(tnode t) {//note root must be passed as argument eg - preorder(tree.root)
		if(t==null) {
			return;
		}else {
			System.out.print(t.data+",");
			preorder(t.left);
			preorder(t.right);
		}
	}
	void postorder(tnode t) {
		if (t==null) {
			return;
		}else {
			postorder(t.left);
			postorder(t.right);
			System.out.print(t.data+",");
		}
	}
	void inorder(tnode t) {
		if (t==null) {
			return;
		}else {
			inorder(t.left);
			System.out.print(t.data+",");
			inorder(t.right);
		}
	}
	void levelorder(tnode t) {
		Vector<tnode> v=new Vector<tnode>();
		v.add(t);
		Vector<Integer> w=new Vector<Integer>();
			while(v.isEmpty()==false) {
				tnode temp=v.firstElement();
				if(temp.left!=null) {
					v.add(temp.left);
				}
				if(temp.right!=null) {
					v.add(temp.right);
				}
				w.add(temp.data);
				v.remove(0);
			}
		int n=w.size();
		for(int i=0;i<n;i++) {
			System.out.print(w.get(i)+" ");
		}
			
	}
	int size(){
		return root.size;
	}
}